modules/ppcp-compat/src/PPEC/class-subscriptionshandler.php

```diff
@@ -0,0 +1,213 @@
+		add_filter( 'wcs_view_subscription_actions', array( $this, 'disable_subscription_actions' ), 999, 2 );
+	}
+
+	/**
+	 * Adds a mock gateway to disguise as PPEC when needed. Hooked onto `woocommerce_payment_gateways`.
+	 * The mock gateway fixes display issues where subscriptions paid via PPEC appear as "via Manual Renewal" and also
+	 * prevents Subscriptions from automatically changing the payment method to "manual" when a subscription is edited.
+	 *
+	 * @param array $gateways List of gateways.
+	 * @return array
+	 */
+	public function add_mock_ppec_gateway( $gateways ) {
+		if ( ! isset( $gateways[ PPECHelper::PPEC_GATEWAY_ID ] ) && $this->should_mock_ppec_gateway() ) {
+			$gateways[ PPECHelper::PPEC_GATEWAY_ID ] = $this->mock_gateway;
+		}
+
+		return $gateways;
+	}
+
+	/**
+	 * Registers BILLING_AGREEMENT as a valid token type for using with the PayPal REST API.
+	 *
+	 * @param array $types List of token types.
+	 * @return array
+	 */
+	public function add_billing_agreement_as_token_type( $types ) {
+		if ( ! in_array( self::BILLING_AGREEMENT_TOKEN_TYPE, $types, true ) ) {
+			$types[] = self::BILLING_AGREEMENT_TOKEN_TYPE;
+		}
+
+		return $types;
+	}
+
+	/**
+	 * Processes subscription renewals on behalf of PayPal Express Checkout.
+	 * Hooked onto `woocommerce_scheduled_subscription_payment_ppec_paypal`.
+	 *
+	 * @param float     $amount The order amount.
+	 * @param \WC_Order $order  The renewal order.
+	 * @return void
+	 */
+	public function process_renewal( $amount, $order ) {
+		add_filter( 'woocommerce_paypal_payments_subscriptions_get_token_for_customer', array( $this, 'use_billing_agreement_as_token' ), 10, 3 );
+
+		$this->ppcp_renewal_handler->renew( $order );
+
+		remove_filter( 'woocommerce_paypal_payments_subscriptions_get_token_for_customer', array( $this, 'use_billing_agreement_as_token' ) );
+	}
+
+	/**
+	 * Short-circuits `RenewalHandler::get_token_for_customer()` to use a Billing Agreement ID for PPEC orders
+	 * instead of vaulted tokens.
+	 *
+	 * @param null|PaymentToken $token    Current token value.
+	 * @param \WC_Customer      $customer Customer object.
+	 * @param \WC_Order         $order    Renewal order.
+	 * @return null|PaymentToken
+	 */
+	public function use_billing_agreement_as_token( $token, $customer, $order ) {
+		if ( PPECHelper::PPEC_GATEWAY_ID === $order->get_payment_method() && wcs_order_contains_renewal( $order ) ) {
+			$billing_agreement_id = $order->get_meta( '_ppec_billing_agreement_id', true );
+
+			if ( $billing_agreement_id ) {
+				$token = new PaymentToken( $billing_agreement_id, 'BILLING_AGREEMENT', new \stdClass() );
+			}
+		}
+
+		return $token;
+	}
+
+	/**
+	 * Limits subscription actions on the frontend when using the mock gateway.
+	 *
+	 * @param array            $actions      Current subscription actions.
+	 * @param \WC_Subscription $subscription Subscription object.
+	 * @return array
+	 */
+	public function disable_subscription_actions( $actions, $subscription ) {
```

Comment:
Hey @jorgeatorres, the docblock says this function is supposed to limit the subscription actions for this mock PPEC gateway but we're just returning the current list of actions?

Is this incomplete or am I missing something? 😅 

Was this supposed to be `return array();`?

Reply:
> Was this supposed to be `return array();`?

Initially, yes, because I wasn't sure if some of the actions would work with the "mock" gateway, but after testing I didn't find any issues. I just forgot to remove the function, but will do that now.